Raise an error for is_signed on quantized types

nairbv opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30527 Raise an error for is_signed on quantized types**

Summary:

When we introduced dtype.is_signed we allowed for support of
quantized types, but we're not sure what the correct result should be.

See discussion at https://github.com/pytorch/pytorch/pull/29511

Differential Revision: [D18765410](https://our.internmc.facebook.com/intern/diff/D18765410)

nairbv committed:
Raise an error for is_signed on quantized types

Summary:

When we introduced dtype.is_signed we allowed for support of
quantized types, but we're not sure what the correct result should be.

See discussion at https://github.com/pytorch/pytorch/pull/29511

[ghstack-poisoned]

albanD commented on torch/csrc/Dtype.cpp:
```diff
@@ -31,6 +33,10 @@ PyObject *THPDtype_is_floating_point(THPDtype *self, PyObject *noargs)
 
 PyObject *THPDtype_is_signed(THPDtype *self, PyObject *noargs)
 {
+  if (at::isQIntType(self->scalar_type)) {
```
Why not add this to the switch in ScalarType.h?
That would give better error messages for cpp users as well.
(AT_ERROR are automatically converted to nice python errors)

nairbv replied:
This also changes ScalarType.h so that it will also raise a C++ exception via the AT_ERROR (got rid of the `toUnderlying(t)` call), but if I allow that to propagate to this call it core dumps. I don't think the python bindings support C++ errors for attribute access, and from python this is an attribute access not a function call:

```
terminate called after throwing an instance of 'c10::Error'
  what():  Unknown ScalarType (isSignedType at ../c10/core/ScalarType.h:322)
frame #0: <unknown function> + 0x6d1c1 (0x7f0ee55dd1c1 in /home/bvaughan/anaconda3/lib/python3.7/site-packages/torch/lib/libc10.so)
frame #1: std::function<std::string ()>::operator()() const + 0x3d (0x7f0eeeadd13d in /home/bvaughan/anaconda3/lib/python3.7/site-packages/torch/lib/libtorch.so)
frame #2: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x28 (0x7f0ee55dc62c in /home/bvaughan/anaconda3/lib/python3.7/site-packages/torch/lib/libc10.so)
frame #3: <unknown function> + 0x9b4069 (0x7f0f0a2b0069 in /home/bvaughan/anaconda3/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #4: THPDtype_is_signed(THPDtype*, _object*) + 0x1f (0x7f0f0a2b0413 in /home/bvaughan/anaconda3/lib/python3.7/site-packages/torch/lib/libtorch_python.so)...
...
Aborted (core dumped)
```

albanD replied:
Is it because the error handling macros are missing from these functions. See [here](https://github.com/pytorch/pytorch/blob/e78b8e0030f7b3f2076688b27c6e0d6b01a98810/torch/csrc/Module.cpp#L195-L201) for example how to use `HANDLE_TH_ERRORS`.

nairbv replied:
I'll try that

gchanan replied:
also put the error message in the exception (so you get a nice message in C++ and the JIT as well), not in the python binding code.

nairbv committed:
Update on "Raise an error for is_signed on quantized types"


Summary:

When we introduced dtype.is_signed we allowed for support of
quantized types, but we're not sure what the correct result should be.

See discussion at https://github.com/pytorch/pytorch/pull/29511

[ghstack-poisoned]

albanD commented on c10/core/ScalarType.h:
```diff
@@ -312,11 +312,14 @@ static inline ScalarType toUnderlying(ScalarType t) {
 
 static inline bool isSignedType(ScalarType t) {
+  if (isQIntType(t)) {
+    TORCH_CHECK(false, "isSignedType not supported for quantized types");
```
nit: `TORCH_CHECK(!isQIntType(t), ...`

nairbv committed:
Update on "Raise an error for is_signed on quantized types"


Summary:

When we introduced dtype.is_signed we allowed for support of
quantized types, but we're not sure what the correct result should be.

See discussion at https://github.com/pytorch/pytorch/pull/29511

[ghstack-poisoned]

nairbv committed:
Update on "Raise an error for is_signed on quantized types"


Summary:

When we introduced dtype.is_signed we allowed for support of
quantized types, but we're not sure what the correct result should be.

See discussion at https://github.com/pytorch/pytorch/pull/29511

Differential Revision: [D18765410](https://our.internmc.facebook.com/intern/diff/D18765410)
